lib/utils/set/set.go

```diff
@@ -0,0 +1,76 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package set
+
+import "golang.org/x/exp/maps"
+
+// Set represents a set data structure, using the map type with an empty value.
+type Set[T comparable] map[T]struct{}
```

Comment:
Hey folks, let's not a push a custom Set structure.

Sets are typically implemented as a valueless map, so in Go we just use a plain `map[Foo]struct{}` when that is necessary (much in the way this is implemented).

Most of the added methods are available as functions in `maps` and others, like Union, can be "implemented" as a `for` if/when needed.

What I'm trying to say is that the language "lacks" Set on purpose, as `map` does its job. Let's not muddle it by adding a custom type to the mix.

Reply:
I guess I'd just like to interrogate this a bit, because there's clearly some interest in not having to rewrite this same code all the time. This isn't even the first halfway implementation I've personally written in Teleport and in any other non-datastructure context I'd appropriately receive review feedback requesting that I didn't duplicate code.

Objectively speaking, most set functionality is _not_ available in `maps`. It provides (with caveats) `Equals` and `Keys` and... nothing else. No set methods are hugely complex in isolation but it is a real drag to keep reimplementing them.

Given that, which of these if any would be more appropriate:
1. A collection of generic set utility functions that operate directly on `map[T]struct{}` in a `sets` package, for example `sets.FromArray`, `sets.ToArray`, `sets.Union`, `sets.Intersection`, `sets.Difference`, `set.IsSubset`, etc. Notably this does not create a new `type Set[...]`.
2. Use of a well-documented and well-tested third-party library providing set functionality, e.g. https://github.com/hashicorp/go-set
3. None of the above. We should always rewrite and retest all set functionality for each use-case.

Reply:
I rarely find that the "entire functionality" of set is necessary. Taking a snippet from the current PR, with changed lines in green:

```diff
+	var currentLogins map[string]struct{}
+	if logins, ok := traits[constants.TraitLogins]; ok {
+		for _, login := range logins {
+			currentLogins[login] = struct{}
+		}
+	}

+	var desiredLogins map[string]struct{}
+	if c.setLogins != "" {
+		for _, login := range splitEntries(c.setLogins) {
+			desiredLogins[login] = struct{}
+		}
+	}

	addLogins := splitEntries(c.addLogins)
+	for _, login := range addLogins {
+		desiredLogins[login] = struct{}
+	}

+	if maps.Equal(currentLogins, desiredLogins) {
		log.Infof("Requested logins match existing, nothing to do: %+v", desiredLogins.ToArray())
		return nil
	}
```

I don't find that much different, or that much more code. We could consider creating a few utilities, or even importing an external library, but I would look for significant need before those either of those.

If you find yourself forced into constant transformations, it's also worth considering whether a different RPC or entity design would mitigate those problems as well.

Reply:
I tend to agree with Alan but I don't have strong feelings.

Would be nice to avoid x/exp/maps though. (We already use depguard to prevent against using x/exp/slices)

Reply:
I... am not sure that I'm totally convinced, but alright, I've unrolled everything and removed all the helper functions.

> Would be nice to avoid x/exp/maps though. (We already use depguard to prevent against using x/exp/slices)

Yep, and as Hugo pointed out, the one function we wanted from it (`Keys()`) is being removed anyway. I've just unrolled that function too.

Reply:
Thanks, Tim. Let me know if this causes frustration again, I'm interested to see the scenarios where it happens. Also let me know if you think I could help.

> the one function we wanted from it (Keys()) is being removed anyway

I think the usual answer is to iterate, like you did.

FWIW I think it's alright to pull "Keys" and "Values" into our codebase somewhere (I would do it in another PR though). I wouldn't be too surprised to see they land on `maps` sometime either, although I'm not sure what the official stance in regards to those is.